File: services/api/src/main/java/com/launchly/testcase/controllers/TestRunController.java
package com.launchly.testcase.controllers;

import com.launchly.common.security.AuthContext;
import com.launchly.testcase.dto.TestRunCaseResponse;
import com.launchly.testcase.dto.TestRunResponse;
import com.launchly.testcase.dto.UpdateTestRunCaseRequest;
import com.launchly.testcase.services.TestService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api")
public class TestRunController {
    private final TestService testService;

    public TestRunController(TestService testService) {
        this.testService = testService;
    }

    @PostMapping("/deployments/{deploymentId}/test-runs")
    public ResponseEntity<TestRunResponse> create(@PathVariable String deploymentId,
                                                   @RequestParam String projectId,
                                                   @RequestParam(required = false) String environmentId) {
        return ResponseEntity.ok(testService.createTestRun(deploymentId, projectId,
                environmentId != null ? environmentId : "", AuthContext.userId()));
    }

    @GetMapping("/test-runs/{id}")
    public ResponseEntity<TestRunResponse> get(@PathVariable String id) {
        return ResponseEntity.ok(testService.getTestRun(id));
    }

    @GetMapping("/test-runs")
    public ResponseEntity<List<TestRunResponse>> list(@RequestParam String projectId) {
        return ResponseEntity.ok(testService.listTestRuns(projectId));
    }

    @GetMapping("/test-runs/{id}/cases")
    public ResponseEntity<List<TestRunCaseResponse>> cases(@PathVariable String id) {
        return ResponseEntity.ok(testService.getTestRunCases(id));
    }

    @PutMapping("/test-runs/{testRunId}/cases/{caseId}")
    public ResponseEntity<TestRunCaseResponse> updateCase(@PathVariable String testRunId,
                                                           @PathVariable String caseId,
                                                           @RequestBody UpdateTestRunCaseRequest request) {
        return ResponseEntity.ok(testService.updateTestRunCase(testRunId, caseId, request, AuthContext.userId()));
    }
}

File: services/api/src/testcase/test-run.controller.ts
import { Controller, Get, Post, Put, Body, Param, Query } from '@nestjs/common';
import { TestService } from './test.service';
import { UpdateTestRunCaseDto } from './dto';
import { CurrentUser, AuthPrincipal } from '../common/decorators/current-user.decorator';
import { Roles } from '../common/decorators/roles.decorator';

@Controller()
export class TestRunController {
  constructor(private readonly testService: TestService) {}

  @Post('deployments/:deploymentId/test-runs')
  @Roles('OWNER', 'ADMIN', 'DEVELOPER', 'TESTER')
  async create(
    @Param('deploymentId') deploymentId: string,
    @Query('projectId') projectId: string,
    @Query('environmentId') environmentId: string | undefined,
    @CurrentUser() user: AuthPrincipal,
  ) {
    return this.testService.createTestRun(deploymentId, projectId, environmentId || '', user.userId);
  }

  @Get('test-runs')
  async list(@Query('projectId') projectId: string) {
    return this.testService.listTestRuns(projectId);
  }

  @Get('test-runs/:id')
  async get(@Param('id') id: string) {
    return this.testService.getTestRun(id);
  }

  @Get('test-runs/:id/cases')
  async cases(@Param('id') id: string) {
    return this.testService.getTestRunCases(id);
  }

  @Put('test-runs/:testRunId/cases/:caseId')
  @Roles('OWNER', 'ADMIN', 'DEVELOPER', 'TESTER')
  async updateCase(
    @Param('testRunId') testRunId: string,
    @Param('caseId') caseId: string,
    @Body() dto: UpdateTestRunCaseDto,
    @CurrentUser() user: AuthPrincipal,
  ) {
    return this.testService.updateTestRunCase(testRunId, caseId, dto, user.userId);
  }
}
